backend.native/tests/framework/values_generics/values_generics.kt

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+// All classes and methods should be used in tests
+@file:Suppress("UNUSED")
+
+package conversions
+
+import kotlin.native.concurrent.isFrozen
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+// Generics
+abstract class BaseData{
+    abstract fun asString():String
+}
+
+data class SomeData(val num:Int = 42):BaseData() {
+    override fun asString(): String = num.toString()
+}
+
+data class SomeOtherData(val str:String):BaseData() {
+    fun anotherFun(){}
+    override fun asString(): String = str
+}
+
+interface NoGeneric<T> {
+  fun myVal():T
+}
+
+data class SomeGeneric<T>(val t:T):NoGeneric<T>{
+  override fun myVal(): T = t
+}
+
+class GenOpen<T:Any?>(val arg:T)
+class GenNonNull<T:Any>(val arg:T)
+
+class GenCollectionsNull<T>(val arg: T, val coll: List<T>)
+class GenCollectionsNonNull<T:Any>(val arg: T, val coll: List<T>)
+
+//Force @class declaration at top of file with Objc variance
+object ForceUse {
+    val gvo = GenVarOut(SomeData())
+}
+
+class GenVarOut<out T:Any>(val arg:T)
+
+class GenVarIn<in T:Any>(tArg:T){
+    private val t = tArg
+
+    fun valString():String = t.toString()
+
+    fun goIn(t:T){
+        //Just taking a val
+    }
+}
+
+class GenVarUse<T:Any>(val arg:T){
+    fun varUse(a:GenVarUse<out T>, b:GenVarUse<in T>){
+        //Should complile but do nothing
+    }
+}
+
+fun variCoType():GenVarOut<BaseData>{
+    val compileVarOutSD:GenVarOut<SomeData> = GenVarOut(SomeData(890))
+    val compileVarOut:GenVarOut<BaseData> = compileVarOutSD
+    return compileVarOut
+}
+
+fun variContraType():GenVarIn<SomeData>{
+    val compileVariIn:GenVarIn<BaseData> = GenVarIn(SomeData(1890))
+    val compileVariInSD:GenVarIn<SomeData> = compileVariIn
+    return compileVariInSD
+}
+
+open class GenBase<T:Any>(val t:T)
+class GenEx<TT:Any, T:Any>(val myT:T, baseT:TT):GenBase<TT>(baseT)
```

Comment:
Would the test pass if `TT` was named `T` instead?
I.e. `class GenEx<T:Any, S:Any>(val myT:S, baseT:T):GenBase<T>(baseT)`

Reply:
I updated test with your example and it's OK.

Reply:
Btw, renaming `T` in `GenEx` to `TT` makes the resulting header invalid.

Reply:
This is related to how the supertype is declared. I wasn't outputting the type arguments, which compiles but creates issues. I'm not sure the best way to get the supertype. I can get the super _class_, but without the type args. During compile, classes are mostly represented by `LazyClassDescriptor`, which has `computeSupertypes()`, which returns `Collection<KotlinType>`, and I can get what I need from there, but that's obviously not a good idea (it's not public). Before I go digging around more, is there an obvious way to get the equivalent with a `ClassDescriptor`?

Reply:
Try `classDescriptor.typeConstructor.supertypes`.

Reply:
Thanks! I didn't see this reply till just now. Spent more time yesterday trying to navigate the parsing code. Assuming I can get this to work, should have an update today.